pkg/tenant/resolver.go

```diff
@@ -0,0 +1,131 @@
+package tenant
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/weaveworks/common/user"
+)
+
+var defaultResolver Resolver = NewSingleResolver()
+
+func DefaultResolver() Resolver {
+	return defaultResolver
+}
+
+func WithDefaultResolver(r Resolver) {
+	defaultResolver = r
+}
```

Comment:
Using a global variable here is not something I am particularly happy about.

I have tried to pass a `tenantResolver` reference around that would itself be instantiated in a Cortex module. But this approach has a massive changeset, as its touching so many areas of the codebase and I decided to go with this approach.

Reply:
I think a global is definitely acceptable in this case. A package could always instantiate it's own resolver down the line if there was a reason to go down that path.